llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Transforms/SplitReduction.cpp

```diff
@@ -0,0 +1,319 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.h"
+#include "iree-dialects/Dialect/LinalgExt/Transforms/Transforms.h"
+#include "iree-dialects/Dialect/LinalgExt/Transforms/Utils.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Arithmetic/IR/Arithmetic.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Linalg/Transforms/Transforms.h"
+#include "mlir/Dialect/SCF/SCF.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/AffineExpr.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/Operation.h"
+#include "mlir/IR/PatternMatch.h"
+#include "llvm/ADT/STLExtras.h"
+
+using namespace mlir;
+using namespace mlir::iree_compiler::IREE::LinalgExt;
+
+namespace {
+
+SmallVector<int64_t> getExpandedShape(ArrayRef<int64_t> inputShape,
+                                      int64_t splitReductionRatio,
+                                      int64_t splitDimParallel) {
+  SmallVector<int64_t> ans;
+  ans.reserve(inputShape.size() + 1);
+  ans.assign(inputShape.begin(), inputShape.end());
+  ans[splitDimParallel] = splitReductionRatio;
+  ans.insert(std::next(ans.begin(), splitDimParallel + 1),
+             inputShape[splitDimParallel] / splitReductionRatio);
+
+  return ans;
+}
+
+SmallVector<int64_t> getCollapsedShape(ArrayRef<int64_t> inputShape,
+                                       int64_t splitReductionRatio, int64_t k,
+                                       int64_t kDimOrig) {
+  SmallVector<int64_t> ans;
+  for (auto elem : inputShape) {
+    ans.push_back(elem);
+  }
+  ans[kDimOrig] = k * splitReductionRatio;
+  return ans;
+}
+
+SmallVector<ReassociationIndices>
+getReassociationIndices(int64_t rank, int64_t splitDimParallel) {
+  SmallVector<ReassociationIndices> reassociationIndices;
+  for (int i = 0; i < rank; ++i) {
+    if (i < splitDimParallel) {
+      reassociationIndices.push_back({i});
+    } else if (i == splitDimParallel) {
+      reassociationIndices.push_back({i, i + 1});
+    } else if (i > splitDimParallel) {
+      reassociationIndices.push_back({i + 1});
+    }
+  }
+  return reassociationIndices;
+}
+
+LogicalResult shouldParallelTopk(iree_compiler::IREE::LinalgExt::TopkOp topkOp,
+                                 PatternRewriter &rewriter, int64_t kDimOrig,
+                                 int64_t splitReductionRatio) {
+  // Determine if we should split the reduction. Requires aligned static shapes
+  // and no input indicies.
+  ShapedType valuesOrigType = topkOp.values().getType().cast<ShapedType>();
+  if (valuesOrigType.getDimSize(kDimOrig) == ShapedType::kDynamicSize) {
+    return rewriter.notifyMatchFailure(topkOp,
+                                       "cannot split dynamic dimension");
+  }
+  if (topkOp.indices()) {
+    return rewriter.notifyMatchFailure(topkOp,
+                                       "input indices aren't supported");
+  }
```

Comment:
Sorry I forget the decision about passing `indices` to topk. Are we going to drop it? Why it's not supported in the PR? I think if we have indices, we don't have to create `linalg.generic` op to offset the indices?

Reply:
the input IR to the Topk op won't have indices, also see the step later after `parallelTopK` is computed, we update the indices with a `linalg.generic` so there is no need for input indices. 

Ultimately we agree the design should change but we're sticking with the current implementation and will redesign once we've gone far enough to see all the ways this can be improved.